pkg/controller/controller.go

```diff
@@ -582,6 +589,21 @@ func (p *csiProvisioner) prepareProvision(ctx context.Context, claim *v1.Persist
 			}
 		}
 	}
+
+	if utilfeature.DefaultFeatureGate.Enabled(features.CrossNamespaceVolumeDataSource) {
+		if claim.Spec.DataSourceRef != nil && claim.Spec.DataSourceRef.Namespace != nil && len(*claim.Spec.DataSourceRef.Namespace) > 0 {
+			switch claim.Spec.DataSourceRef.Kind {
+			case snapshotKind:
+				if *(claim.Spec.DataSourceRef.APIGroup) != snapshotAPIGroup {
+					return nil, controller.ProvisioningFinished, fmt.Errorf("the PVC source does not belong to the right APIGroup. Expected %s, Got %s", snapshotAPIGroup, *(claim.Spec.DataSourceRef.APIGroup))
+				}
+				rc.snapshot = true
+			default:
+				return nil, controller.ProvisioningFinished, fmt.Errorf("the PVC source specified via DataSourceRef with non-empty namespace does not belong to the right Kind.  Expected %s, Got %s", snapshotKind, claim.Spec.DataSourceRef.Kind)
```

Comment:
Are we allowing pvc source and data populators too?

Reply:
Yes. Added. https://github.com/kubernetes-csi/external-provisioner/commit/1f9df93e646bbd126bf7bcc172a7e2a3a2da4e8e